#### Dragonfly Statistics youtube video notes: start with intercept only model, using 1. no eplanatory variables in this model. should just turn out to be the mean of mpg in mtcars. start adding regression coefficients in, but only if they're useful and if they make model better. you have to set out how far you have to go, because you probably don't need to use all variables. must specify how big model will be, some limitations.  do that with 'scope=formula(Fitall)'. that's as big as the model will go and no further. 

#### output is a ranking. law of parsimony and multicollinearity important to consider. best option for modelis to add wt. AIC will drop ifyou add wt. a drop in AIC is improvement in the model. nine more scenarios we can add in variables. add in one by one and reappraise AIC value. keep an eye on none in carats where it moves.when none is top, that's a stopping condition. of the ten, best outcome or subset of the variables is wt, cyl and hp. earlier with backward elimination is different set of variables. AIC is different as well. backward gave better answer than forward selection. 

#### R square is the percentage of variation in the DV explained by the IV. 

### Create original model

In [1]:
fitstart = lm(mpg~1,data=mtcars)

In [2]:
summary(fitstart)


Call:
lm(formula = mpg ~ 1, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-9.6906 -4.6656 -0.8906  2.7094 13.8094 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   20.091      1.065   18.86   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 6.027 on 31 degrees of freedom


#### Notice that the estimate for the intercept is 20.091, which is simply the average mpg for the 28 vehicles in the original mtcars dataset.

In [4]:
Fitall <- lm(mpg~.,data=mtcars)

In [5]:
step(fitstart, direction = "forward", scope=(formula(Fitall)))

Start:  AIC=115.94
mpg ~ 1

       Df Sum of Sq     RSS     AIC
+ wt    1    847.73  278.32  73.217
+ cyl   1    817.71  308.33  76.494
+ disp  1    808.89  317.16  77.397
+ hp    1    678.37  447.67  88.427
+ drat  1    522.48  603.57  97.988
+ vs    1    496.53  629.52  99.335
+ am    1    405.15  720.90 103.672
+ carb  1    341.78  784.27 106.369
+ gear  1    259.75  866.30 109.552
+ qsec  1    197.39  928.66 111.776
<none>              1126.05 115.943

Step:  AIC=73.22
mpg ~ wt

       Df Sum of Sq    RSS    AIC
+ cyl   1    87.150 191.17 63.198
+ hp    1    83.274 195.05 63.840
+ qsec  1    82.858 195.46 63.908
+ vs    1    54.228 224.09 68.283
+ carb  1    44.602 233.72 69.628
+ disp  1    31.639 246.68 71.356
<none>              278.32 73.217
+ drat  1     9.081 269.24 74.156
+ gear  1     1.137 277.19 75.086
+ am    1     0.002 278.32 75.217

Step:  AIC=63.2
mpg ~ wt + cyl

       Df Sum of Sq    RSS    AIC
+ hp    1   14.5514 176.62 62.665
+ carb  1   13.7724 177.40 62.805
<no


Call:
lm(formula = mpg ~ wt + cyl + hp, data = mtcars)

Coefficients:
(Intercept)           wt          cyl           hp  
   38.75179     -3.16697     -0.94162     -0.01804  


#### with a model including wt, cyl, and hp, you are done. There is no point in adding any more terms, because according to the table, the AIC will actually increase if you do. Do the same thing you did for the backward elimination model: build a model that only contains wt, cyl, and hp. You called it fitsome in the backward elimination model; call this one fitsome2:

In [6]:
fitsome <- lm(mpg~wt+cyl+hp, data=mtcars)

In [7]:
summary(fitsome)


Call:
lm(formula = mpg ~ wt + cyl + hp, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.9290 -1.5598 -0.5311  1.1850  5.8986 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 38.75179    1.78686  21.687  < 2e-16 ***
wt          -3.16697    0.74058  -4.276 0.000199 ***
cyl         -0.94162    0.55092  -1.709 0.098480 .  
hp          -0.01804    0.01188  -1.519 0.140015    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.512 on 28 degrees of freedom
Multiple R-squared:  0.8431,	Adjusted R-squared:  0.8263 
F-statistic: 50.17 on 3 and 28 DF,  p-value: 2.184e-11


#### When you did the backward elimination model, you compared the initial model (all ten predictors) with the optimized model, and concluded that the models are similar as far as R2 is concerned, so the model with just 3 predictors seems preferable based upon simplicity alone. For the forward selection model, there really is no starting point sort of model. You just used the average mpg to get a starting point without any predictor variables.

Nonetheless, take a look at the specifics of the forward selection model shown above:

The multiple R2 = 0.8431. The practical interpretation of this is that the model explains 84.31% of the variation in the mpg variable, and there is another 15.69% of the variation that can be chalked up to noise or random error.
There is an adjusted R2 = 0.8263. The "adjustment" is a modification that is supposed to take into account the number of terms in the model. For your purposes, you are more interested in the adjusted R2 than the multiple R2.
There is an F-statistic (50.17 with 3 and 28 degrees of freedom) and a p value of 0.00000000002184 (this is represented in scientific notation in R) which indicates that the model is better than no model at all, because the p value is less than 0.05.
The R2 for the backward elimination model and the forward selection model are similar. So, which one is better? One could argue that the AIC for the backward elimination model is slightly lower, so it must be better. But for all practical purposes, they are essentially the same. However, did you notice the terms in the predictor model are not the same? That is actually a pretty common result. The truth is that in many cases, there might not just be a single 'best' model that is far superior to the 'second best' model, but there might be a cluster of models that are essentially equally good.



## Hybrid Stepwise - Forward and Backward Selection

#### video notes  fwd selection is intercept only. response variables get added one at a time only if they improve the model. change forward to both.same scope.  best variable to add to the model is lowest AIC.note plus and minus signs in output. minus is it can get taken out again. model might realize it will make model better. multicollinearity and law of parsimony.at the top of the list might be a minus, as in remove some variable that was added previously and variable selection procedure thinks it is time to take it out.  

#### model is similar to forward selection model. 

In [8]:
step(fitstart, direction = "both", scope=formula(Fitall))

Start:  AIC=115.94
mpg ~ 1

       Df Sum of Sq     RSS     AIC
+ wt    1    847.73  278.32  73.217
+ cyl   1    817.71  308.33  76.494
+ disp  1    808.89  317.16  77.397
+ hp    1    678.37  447.67  88.427
+ drat  1    522.48  603.57  97.988
+ vs    1    496.53  629.52  99.335
+ am    1    405.15  720.90 103.672
+ carb  1    341.78  784.27 106.369
+ gear  1    259.75  866.30 109.552
+ qsec  1    197.39  928.66 111.776
<none>              1126.05 115.943

Step:  AIC=73.22
mpg ~ wt

       Df Sum of Sq     RSS     AIC
+ cyl   1     87.15  191.17  63.198
+ hp    1     83.27  195.05  63.840
+ qsec  1     82.86  195.46  63.908
+ vs    1     54.23  224.09  68.283
+ carb  1     44.60  233.72  69.628
+ disp  1     31.64  246.68  71.356
<none>               278.32  73.217
+ drat  1      9.08  269.24  74.156
+ gear  1      1.14  277.19  75.086
+ am    1      0.00  278.32  75.217
- wt    1    847.73 1126.05 115.943

Step:  AIC=63.2
mpg ~ wt + cyl

       Df Sum of Sq    RSS    AIC
+ hp    1    


Call:
lm(formula = mpg ~ wt + cyl + hp, data = mtcars)

Coefficients:
(Intercept)           wt          cyl           hp  
   38.75179     -3.16697     -0.94162     -0.01804  


#### Hierarchical Regression
In hierarchical regression, you get to pick the variables that are being added or removed next, which allows you to make statements about how much variance is added "over and above" other variables. This approach can be quite useful especially when answering stakeholder questions about what variables are most important.

Summary
Stepwise regression has three basic approaches - backward elimination, forward selection, and a combination of the two.
Backward elimination creates a model with all predictor variables included, and then removes them one at a time until the model is optimized.
Forward selection starts with no predictor variables included, and adds them one at a time until the model is optimized.
The combination approach starts with no predictor terms. Every time a term is either added or removed from the model, it is compared with the quality of all other models that either add a single predictor or remove a single predictor until the model is optimized.

#### Key Terms
Below is a list and short description of the important keywords learned in this lesson. Please read through and go back and review any concepts you do not fully understand. Great Work!

Keyword	Description
Stepwise Regression 	A process to determine exactly what variables make up the best-fitting regression model.

Multiple Regression	 A regression with multiple independent variables.

Overfitting	 When a model fits the data so well that it won't fit future data.

Principle of Parsimony	When all other things are equal, go with the simplest explanation.

Backward Elimination	Starting with all the predictors and removing them one at a time to find the best-fitting model.

Forward Selection	Starting with no predictors and adding them one at a time to find the best-fitting model.

Hybrid Stepwise Regression	Starting with no predictors and adding or removing them one at a time to find the best-fitting model.

Akaike Information Criteria (AIC)	An indicator of model fit quality. The smaller, the better.


Key R Code
Keyword	Description
step()	A function to perform stepwise regression.

direction="backward"	An argument to step() that performs backward elimination.

direction="forward"	An argument to step() that performs forward selection.

scope=	An argument to step() that specifies what model you are moving towards.

direction="both"	An argument to step() that specifies hybrid stepwise regression.